server/data/data-jmap/src/test/scala/org/apache/james/jmap/api/identity/IdentityRepositoryTest.scala

```diff
@@ -0,0 +1,145 @@
+  var identityFactory: DefaultIdentitySupplier = _
+  var customIdentityDAO: CustomIdentityDAO = _
+
+
+  @BeforeEach
+  def setUp(): Unit = {
+    customIdentityDAO = new MemoryCustomIdentityDAO()
+    identityFactory = mock(classOf[DefaultIdentitySupplier])
+    testee = new IdentityRepository(customIdentityDAO, identityFactory)
+  }
+
+  @Test
+  def saveShouldSuccess(): Unit = {
+    when(identityFactory.userCanSendFrom(any(), any())).thenReturn(true)
+    assertThatCode(() => SMono.fromPublisher(testee.save(BOB, CREATION_REQUEST)).block())
+      .doesNotThrowAnyException()
+  }
+
+  @Test
+  def saveShouldFailWhenUserCanNotSendFrom(): Unit = {
+    when(identityFactory.userCanSendFrom(any(), any())).thenReturn(false)
+    assertThatThrownBy(() => SMono.fromPublisher(testee.save(BOB, CREATION_REQUEST)).block())
+      .isInstanceOf(classOf[ForbiddenSendFromException])
+  }
+
+  @Test
+  def listShouldReturnCustomAndServerSetEntries(): Unit = {
+    SMono.fromPublisher(customIdentityDAO.save(BOB, CREATION_REQUEST)).block()
+    when(identityFactory.listIdentities(BOB)).thenReturn(List(IDENTITY1))
+
+    assertThat(SFlux.fromPublisher(testee.list(BOB)).collectSeq().block().asJava)
+      .containsExactlyInAnyOrder(IDENTITY1, SFlux(customIdentityDAO.list(BOB)).blockFirst().get)
+  }
+
+  @Test
+  def listShouldReturnCustomEntryWhenIdExistsInBothCustomAndServerSet(): Unit = {
+    SMono.fromPublisher(customIdentityDAO.save(BOB, CREATION_REQUEST)).block()
+    val identity: Identity = SFlux(customIdentityDAO.list(BOB)).blockFirst().get
+    val differentIdentityWithSameId: Identity = identity.copy(name = IdentityName("different name"))
+    when(identityFactory.listIdentities(BOB)).thenReturn(List(differentIdentityWithSameId))
+
+    assertThat(SFlux.fromPublisher(testee.list(BOB)).collectSeq().block().asJava)
+      .containsExactlyInAnyOrder(identity)
+  }
+
+  @Test
+  def updateShouldFailWhenIdNotFoundInBothCustomAndServerSetDAO(): Unit = {
+    when(identityFactory.listIdentities(BOB)).thenReturn(List())
+
+    assertThatThrownBy(() => SMono.fromPublisher(testee.update(BOB, IdentityId.generate, UPDATE_REQUEST)).block())
+      .isInstanceOf(classOf[IdentityNotFoundException])
+  }
+
+  @Test
+  def updateShouldSuccessWhenCustomNotFoundAndServerSetExists(): Unit = {
+    when(identityFactory.listIdentities(BOB)).thenReturn(List(IDENTITY1))
+
+    assertThatCode(() => SMono.fromPublisher(testee.update(BOB, IDENTITY1.id, UPDATE_REQUEST)).block())
+      .doesNotThrowAnyException()
+  }
+
+  @Test
+  def updateShouldSuccessWhenCustomExists(): Unit = {
+    when(identityFactory.listIdentities(BOB)).thenReturn(List())
+    SMono.fromPublisher(customIdentityDAO.save(BOB, CREATION_REQUEST)).block()
+    val identity: Identity = SFlux(customIdentityDAO.list(BOB)).blockFirst().get
+
+    assertThatCode(() => SMono.fromPublisher(testee.update(BOB, identity.id, UPDATE_REQUEST)).block())
+      .doesNotThrowAnyException()
+  }
```

Comment:
What are the results of these updates?

Can I update a server set identity twice?